MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs

```diff
@@ -4980,22 +4995,40 @@ public string GetHwTonemapFilter(EncodingOptions options, string hwTonemapSuffix
             var newfilters = new List<string>();
             var noOverlay = swFilterChain.OverlayFilters.Count == 0;
             var supportsHwDeint = _mediaEncoder.SupportsFilter("yadif_videotoolbox");
+            var supportsHwScale = _mediaEncoder.SupportsFilter("scale_vt");
+            // VideoToolbox is special. It does not use a separate tone mapping filter like others. Instead, it performs both tone mapping and scaling in a single filter.
+            var useHwToneMapping = IsVideoToolboxTonemapAvailable(state, options) && supportsHwScale;
             // fallback to software filters if we are using filters not supported by hardware yet.
-            var useHardwareFilters = noOverlay && (!doDeintH2645 || supportsHwDeint);
+            var useHardwareFilters = noOverlay && (!doDeintH2645 || supportsHwDeint || supportsHwScale);
```

Comment:
So you don’t want users to enable tone-mapping and subtitle burn-in at the same time? It does result in a more complex filter pipeline, but I know many users want it.

Reply:
This decision was made a long time ago when the only hardware filter available was yadif.

It serves as a fallback to the safest software chain, as the hardware filter currently does not handle overlay.

However, our software filter chain also cannot perform tone mapping. So, to enable tone mapping functionality with subtitles burn-in, I propose do an `hwdownload` to yuv420 and attaching the software subtitle filters to the chain. Would this approach be acceptable? Or do I have to do something more.

Reply:
We are not in a hurry, I would rather have a complete hardware pipeline. I'm actually more interested in whether Apple's fixed-function hardware supports overlay/image blending, or we can achieve through a Metal kernel.

If you are interested in it, we can try writing an `overlay_vt/videotoolbox` filter for ffmpeg (refer to [`overlay_cuda.cu`](https://github.com/FFmpeg/FFmpeg/blob/master/libavfilter/vf_overlay_cuda.cu) and port it to `overlay_videotoolbox.metal`, we already have a [`yadif_videotoolbox.metal`](https://github.com/FFmpeg/FFmpeg/blob/master/libavfilter/metal/vf_yadif_videotoolbox.metal) that can be used as a template). I don't have Apple hardware but can provide you with technical support in writing code for FFmpeg. Once completed, submit it to upstream FFmpeg and everything will be perfect.

Reply:
I'm currently working on a Metal implementation of the overlay filter. Apple's VideoToolbox does not expose any image blending function, so I'm required to use a custom Metal kernel. I already have a working kernel that performs basic alpha compositing and positioning, and I'm in the process of porting it to ffmpeg.

In examining the current usage pattern in Jellyfin, I've observed that most hardware overlay filters are used with an additional custom `alphaSrc` filter. I'm curious about the purpose of this filter. Perhaps I can adjust my implementation to eliminate the need for it and instead let the overlay filter to handle the issue more effectively.

Reply:
You can't just consider PGS graphical subtitles. `alphasrc` is used in conjunction with the subtitles filter to convert SSA/ASS/SRT text subtitles to images. This is unavoidable.

Reply:
I have made some progress and the filter produces outputs(although not correct) in some cases. A big problem I found is that we cannot `hwupload` arbitrary image because the `HWContext` of `VideoToolBox` does not support common colorspaces seen in a lot of images like `rgba` used by pngs. Is this normal for HW overlay filters and could I do anything to convert that to something `VideoToolBox` could recognize(like yuv)?

Reply:
The formats that can be used in HW filters are usually restricted in the [hwcontext_videotoolbox](https://github.com/FFmpeg/FFmpeg/blob/0895ef0d6d6406ee6cd158fc4d47d80f201b8e9c/libavutil/hwcontext_videotoolbox.c#L48).

For the subtitle burn-in use case, we need `main (nv12/yuv420p) + overlay (bgra/rgba/yuva420p)`, which means that the `overlay` input must contain at least an alpha channel.

Also, since the `scale_vt` filter (fixed-function HW) is not capable of pixel format conversion, you may need to add a [RGB2YUV/YUV2RGB](https://github.com/FFmpeg/FFmpeg/blob/0895ef0d6d6406ee6cd158fc4d47d80f201b8e9c/libavfilter/opencl/colorspace_common.cl#L137) function in the Metal kernel of `overlay_vt` to handle this.

Reply:
It seems Apple [supports](https://developer.apple.com/documentation/corevideo/kcvpixelformattype_32rgba) `rgba` so you can add it. The difference between it and `bgra` is the channel order. Note that PGS subtitles natively use `bgra`.

If I understand correctly, [VTPixelTransferSessionTransferImage](https://developer.apple.com/documentation/videotoolbox/1503548-vtpixeltransfersessiontransferim) allows you to `convert` pixel formats. Assuming the vt encoders also support `bgra`. Then `overlay_vt` can be simplified to (bgra+bgra).

```
// pseudo filter pipeline
[0:s]hwupload[sub] // bgra, from pgssub or alphasrc
[0:v]scale_vt=format=bgra[main] // add convert support for `scale_vt`
[main][sub]overlay_vt[out] //bgra to encoder
```

Reply:
I *think* I got it. That method is calling `CVMetalTextureCacheCreateTextureFromImage` with width and height calculated from `CVPixelBufferGetWidthOfPlane`, but RGBA buffers are not planer, so that method would return 0, hence the `kCVReturnInvalidArgument`. I think I need to fix that first.

Reply:
Only the video decoder generates yuv interlaced videos, so the author of `yadif_vt` probably only tested planar formats.